### PR TITLE
Expand owners franchise

### PR DESCRIPTION
**Pull request by calebamiles**
**What this PR does / why we need it**:

Expanding the group of people who are able to review PRs is important for the growing community of Kubernetes contributors. This change uses git commit history to calculate a list of possible reviewers for various parts of the project. 

**Which issue this PR fixes**
 `fixes #1422,#1389` 

**Release note**

The list of reviewers to OWNERS files has been updated using tooling [contributed by CoreOS](https://github.com/coreos/findowner) Thanks @hongchaodeng! 

**Overview**

This patch adds a set of reviewers to OWNERS files across various
packages in Kubernetes. The big idea is to maintain this list of people
who have worked on parts of the Kubernetes code base who can potentially
help with code reviews. This list will be consumed by bots that will
automatically ask people to review code and assign PRs for review.

Discussions:
- https://github.com/kubernetes/contrib/issues/1422
- https://github.com/kubernetes/contrib/issues/1389
- https://groups.google.com/forum/#!topic/kubernetes-wg-contribex/EEEyMisp0M4

**Selected Reviewers**

There are approximately 270 people identified for the role of pkg level
reviewer based on their work in various packages. Each of these people will be contacted via the email associated with their account in order to confirm their participation as a reviewer
Of course if anyone wishes to opt out immediately they can comment on this PR:

You may opt-out in two different ways:
1. Remove permanently; your GitHub handle will be added to a blacklist
2. Remove from a single package; your GitHub handle will be added to a
   pkg blacklist and removed from this patch

Please comment on this issue if you want to opt-out. See the patch for
the pkgs you are in.

**Implementation**

The basic algorithm:
- Walk directories of "pkg" and gather up commit history
- It selects top N (default is 3) committers for each directory. If there are more than N people having the same number of commits, it outputs them all.
- It ranks people (identified by GitHub handle) by the number of commits they made.
- The commits are pulled from past two years history excluding "merge" commits.

https://github.com/coreos/findowner

_Implementation Notes_
- It has a depth limit (default is 3). For example, given depth limit 3, "./pkg/apis/apps" is allowed, while "./pkg/apis/apps/install" isn't.
- It has an excluded list. See "./main.go" excludedDirList variable.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31752)

### CODE REVIEW

**Comment by foxish** [comment]
Does the wider list of reviewers contain folks who are not members of the kubernetes organization? If that is the case, we may need to prune the list (and invite them into the organization) before auto-assignment can work correctly with them.

**Comment by philips** [comment]
@foxish Why would we prune the list? Wouldn't we just add all of the folks who are not members to a read-only team?

**Comment by foxish** [comment]
@philips I meant temporarily removing them, and adding them after they all accept their invites. It might take some time for folks to accept the invite. https://github.com/kubernetes/kubernetes.github.io/pull/952 is what we did on the docs repo for example. If we let them stay in the lists without having accepted the invitation, the bot that does auto-assignment thrashes about and tries to repeatedly assign them, and eventually triggers github's abuse detection mechanisms IIRC. 

Also, @calebamiles Would it be possible to alphabetize the lists of reviewers and approvers that are generated as well?

**Comment by calebamiles** [comment]
It would be possible to alphabetize the list @foxish. Unfortunately if a contributor has not made their membership in the Kubernetes organization public there doesn't seem to be a good way to know that they're part of the Organization, at least with my permission set. I don't know if it's entirely reasonable to force people to publicize their membership in an organization.

**Comment by foxish** [comment]
@calebamiles Sent you an invite to add you to the org. With that, you should be able to use https://developer.github.com/v3/orgs/members/#check-membership to check people's membership in the org with that.

File: pkg/api/rest/OWNERS [hunk 1]
```diff
@@ -3,3 +3,4 @@ reviewers:
 - wojtek-t
 - thockin
 - caesarxuchao
+- smarterclayton
```
Comment (smarterclayton) [inline pkg/api/rest/OWNERS]: Part of the the reason for this was to catch the original authors, which this looks like it has done for a number of important core packages that have been around for longer than 1 year.

**Comment by philips** [comment]
On the community hangout call I heard three pieces of feedback from @brendandburns and @smarterclayton 

- [ ] Make this opt-in instead of opt-out (Implementation: Email + Google Form probably)
- [ ] Target "active" participants (Implementation: Active in comments or commits for kubernetes.git in the last two releases?)
- [ ] Reduce target audience to ~100 people (Implementation: opt-in and "active" filter may be sufficient, if not reduce pkg depth or sort by pkg size)

**Comment by smarterclayton** [comment]
SGTM - identify reviewers by age (current mechanism) and then filter to
active contributors.

**Comment by calebamiles** [comment]
I would like to propose a slight deviation from the heuristic discussed at the community meeting. I liked the idea that @foxish had to limit the initial pool of new reviewers to people who are currently members of the Kubernetes GitHub organization because it's fairly easy for me to compute in code whereas looking at the candidate list and then seeing how active they are against the Kubernetes repo would be somewhat more difficult for me to compute. Using the heuristic of `member of organization = active` we would add **95** new reviewers as follows

```json
[
 "@AdoHe",
 "@Q-Lee",
 "@Random-Liu",
 "@a-robinson",
 "@aaronlevy",
 "@aledbf",
 "@andreykurilin",
 "@asalkeld",
 "@ashcrow",
 "@aveshagarwal",
 "@bparees",
 "@bprashanth",
 "@bryk",
 "@caseydavenport",
 "@childsb",
 "@ciwang",
 "@cjcullen",
 "@danwinship",
 "@david-mcmahon",
 "@dcbw",
 "@deads2k",
 "@dims",
 "@dshulyak",
 "@ericchiang",
 "@euank",
 "@fabianofranz",
 "@fabioy",
 "@feihujiang",
 "@feiskyer",
 "@fgrzadkowski",
 "@foxish",
 "@freehan",
 "@ghodss",
 "@girishkalele",
 "@gmarek",
 "@goltermann",
 "@gtank",
 "@hodovska",
 "@hongchaodeng",
 "@hurf",
 "@hw-qiaolei",
 "@ingvagabund",
 "@jackgr",
 "@janetkuo",
 "@jayunit100",
 "@jbeda",
 "@jfrazelle",
 "@jiangyaoguo",
 "@jimmidyson",
 "@jingxu97",
 "@jlowdermilk",
 "@jsafrane",
 "@jszczepkowski",
 "@justinsb",
 "@jwforres",
 "@kargakis",
 "@kevin-wangzefeng",
 "@lukaszo",
 "@luxas",
 "@magicwang-cn",
 "@markturansky",
 "@marun",
 "@matchstick",
 "@mbohlool",
 "@mfojtik",
 "@mksalawa",
 "@mqliang",
 "@mtaufen",
 "@mwielgus",
 "@ncdc",
 "@nhlfr",
 "@pigmej",
 "@piosz",
 "@pmorie",
 "@preillyme",
 "@pweil-",
 "@rajatchopra",
 "@resouer",
 "@ronnielai",
 "@rootfs",
 "@saad-ali",
 "@simonswine",
 "@sjenning",
 "@sjpotter",
 "@soltysh",
 "@sttts",
 "@swagiaal",
 "@therc",
 "@timothysc",
 "@timstclair",
 "@tmrts",
 "@uruddarraju",
 "@vmarmol",
 "@xiang90",
 "@yifan-gu"
]
```

new is defined here as not appearing in any previous `OWNERS` file. If this is ok I can push the pending commit to my branch and we can talk about adjusting the heuristic for determining activeness in another effort. Does that sound reasonable @smarterclayton @brendandburns?

**Comment by philips** [comment]
Caleb and I talked in person. I think starting with the people who are already organization members is fine in the initial pass.

But, we will also look into using bigquery to get the "active" users as outlined in https://github.com/kubernetes/kubernetes/pull/31752#issuecomment-244364574 too

**Comment by foxish** [comment]
Before this change is merged, we would need our bot to start picking up the `reviewers` from the OWNERS files instead of `assignees` as it currently does. I think @apelisse was working on this change.

**Comment by philips** [comment]
@foxish OK, We likely have 7+ days before we make the changes in any case.

**Comment by grodrigues3** [comment]
@apelisse 's pull request to read `approvers` and `reviewers` instead of `assignees`: https://github.com/kubernetes/contrib/pull/1627

**Comment by lavalamp** [comment]
Finally going through my review queue. What's the status of this PR? Can we split it into multiple smaller PRs?

**Comment by philips** [comment]
Status: @lavalamp We are rather blocked on this discussion https://github.com/kubernetes/kubernetes/pull/32096 and we were asked to make this a single PR to focus discussion.

**Comment by lavalamp** [comment]
OK. Ping me when this is ready for a look.

**Comment by calebamiles** [comment]
I'd love for you to take a look when you get a chance @lavalamp. Emails are waiting to go [in Google Drive](https://drive.google.com/open?id=0B72FHdA0LUB8cTBnMUlBbThQS2s). I'll be sending them out as a Google Group invitation so that the spam filters don't go wild on my work email address.

**Comment by lavalamp** [comment]
Emails look good to me.

**Comment by calebamiles** [comment]
Great @lavalamp, and now this PR should actually merge in without conflicts :). So TL;DR we have 138 new reviewers and as soon as I've gotten my Google Group I'll go ahead and send out those emails as listed in the drive folder.

**Comment by bgrant0607** [comment]
@philips @smarterclayton @brendandburns Now that 1.4 is out, I would like to just merge this and then iterate on the exact contents of the individual OWNERS files. In many cases, we would have manually dug through history to figure out who should review changes.

Note that approvers doesn't gate PR merges yet.

**Comment by errm** [comment]
so I got an email from @calebamiles, or a bot more likely... it said to add a comment here to say I was ok with being a reviewer...
... so I am ok with that ...

**Comment by abhgupta** [comment]
Commenting here in response to the reviewer preferences/invite email - I would like to opt in to the pull request review process for all paths identified.

**Comment by ivan4th** [comment]
(responding to the review invitation)
I accept the invitation for all paths listed in the e-mail.

**Comment by jayunit100** [comment]
I also accept

**Comment by markturansky** [comment]
reviewer status = on

**Comment by mfanjie** [comment]
I received invitation letter for k8s reviewers, And I'd like to opt into all the paths identified.

**Comment by resouer** [comment]
I also accept

**Comment by ping035627** [comment]
I'd like to opt into pkg/genericapiserver/validation - integration/scheduler, thanks for the invitation.

**Comment by lixiaobing10051267** [comment]
I also accept, thanks very much, and It is better for me to review the following paths:
1. docs/devel
2. docs/getting-started-guides
3. docs/proposals
4. docls/user-guide
5. pkg/apiserver
6. plugin/cmd/kube-scheduler

**Comment by bmoylan** [comment]
I accept for `cluster/centos`

**Comment by 249043822** [comment]
I'd like to opt into pkg/quota/evaluator

**Comment by jimmidyson** [comment]
Accepted - thanks!

**Comment by anish** [comment]
Opt in for pkg/volume/iscsi

**Comment by gmarek** [comment]
I accept

**Comment by sttts** [comment]
I also accept.

**Comment by wojtek-t** [comment]
I accept

**Comment by nebril** [comment]
I accept

**Comment by PiotrProkop** [comment]
I accept.

**Comment by mbohlool** [comment]
I accept.

**Comment by dims** [comment]
I accept

**Comment by nhlfr** [comment]
I accept

**Comment by porridge** [comment]
I accept.

**Comment by screeley44** [comment]
I accept and would also like to be considered for review of the following paths:
  
Opt in for  /pkg/volume
Opt in for  /pkg/controller/volume

**Comment by fabianofranz** [comment]
I also accept for all paths identified, specially `pkg/kubectl`.

**Comment by mbruzek** [comment]
I accept the invitation, and would like to opt-in. Thank you.

**Comment by matthewdupre** [comment]
I accept for cluster/addons/calico-policy-controller

**Comment by ymqytw** [comment]
I accept

**Comment by piosz** [comment]
I accept

Opt in:
cluster/addons
cluster/addons/cluster-monitoring
cluster/addons/fluentd-elasticsearch
cluster/addons/fluentd-gcp/fluentd-gcp-image
pkg/apis/autoscaling
pkg/controller/podautoscaler

Opt out:
test/kubemark
test
test/kubemark/resources
cluster/addons/dashboard
pkg/fieldpath
test/e2e

**Comment by guangxuli** [comment]
I also accept the paths identified.

**Comment by feiskyer** [comment]
I accept

**Comment by dalanlan** [comment]
I accept

**Comment by lojies** [comment]
I accept

**Comment by anish** [comment]
Accept, Opt in for `pkg/volume/iscsi`

**Comment by Starefossen** [comment]
I accept ✨

**Comment by Pensu** [comment]
I accept.

**Comment by m1093782566** [comment]
I accept

**Comment by soltysh** [comment]
I accept:

Opt in:
pkg/api/testing
pkg/apis/batch
pkg/apiserver/audit
pkg/controller/job
pkg/controller/scheduledjob
pkg/kubectl

**Comment by aledbf** [comment]
I accept

**Comment by taimir** [comment]
I accept

**Comment by childsb** [comment]
I accept

**Comment by lixiaobing10051267** [comment]
I accept, thanks, besides the automated "pkg/credentialprovider/aws", the followings can also be considered:
1. docs/devel
2. plugin/pkg/scheduler/algorithm

**Comment by huang195** [comment]
I accept

**Comment by hongchaodeng** [comment]
I accept.

opt in:
- pkg/fields
- pkg/labels
- pkg/selection
- pkg/storage
- pkg/storage/etcd
- pkg/storage/etcd3
- pkg/storage/storagebackend

Thanks!

**Comment by ursuad** [comment]
In response to the reviewer preferences/invite email - I would like to opt in to the pull request review process for all paths identified.

**Comment by bryk** [comment]
I opt in.

**Comment by jianhuiz** [comment]
I accept.

**Comment by douglas-gibbons** [comment]
I accept.

Opt in:
*  cluster/addons/addon-manager

**Comment by jayunit100** [comment]
I accept.

File: cmd/genswaggertypedocs/OWNERS [hunk 2]
```diff
@@ -0,0 +1,3 @@
+reviewers:
+- david-mcmahon
+- luxas
```
Comment (luxas) [inline cmd/genswaggertypedocs/OWNERS]: Can I opt out of this one please?
Don't know anything really about the swagger generation process

**Comment by luxas** [comment]
I accept

Summary for me:
Opt-in:
- build/build-image
- build/build-image/cross
- build/debian-iptables
- build/pause
- cluster/addons/addon-manager
- cluster/addons/dashboard
- cluster/images
- cluster/images/etcd
- cluster/images/flannel
- cluster/images/hyperkube
- cmd/kubeadm
- cmd/kubeadm/app
- docs/getting-started-guides

Opt-out:
- cmd/genswaggertypedocs
- pkg/api/service
- pkg/util/flowcontrol
- pkg/util/keymutex
- test/e2e/generated

**Comment by mtaufen** [comment]
I accept

Summary for me:
Opt-in as reviewer:
- pkg/kubelet
- cmd/kubelet

Opt-out:
- pkg/kubemark
- pkg/util/config

**Comment by adityakali** [comment]
I accept:
 - cluster/gce/gci

**Comment by asalkeld** [comment]
I accept

**Comment by chrislovecnm** [comment]
I accept

**Comment by zhouhaibing089** [comment]
I accept

**Comment by hodovska** [comment]
I accept

Opt-in:
- pkg/admission
- test/integration/replicationcontroller

**Comment by ingvagabund** [comment]
I accept

Opt-in:
* test/images/mount-tester
* plugin/pkg/scheduler

Opt-out:
* pkg/kubelet/cm

**Comment by chakri-nelluri** [comment]
I accept

Opt-in: 
- pkg/volume/flexvolume

**Comment by joe2far** [comment]
I accept

**Comment by calebamiles** [comment]
@k8s-bot verify test this

**Comment by grodrigues3** [comment]
@calebamiles I think some tests are failing because of #34572, which may be failing by trying to diff owners' files in directories with generated file.  For example: 

```/go/src/k8s.io/kubernetes/_tmp/api-reference/autoscaling/OWNERS```

I think we can just delete those owners files

**Comment by nikhiljindal** [comment]
Yes removing OWNERS from directories with only auto generated code seems good to me.
No one can directly change the code in those directories. They will have to modify some other code that should have the appropriate OWNERS file.

**Comment by xiangpengzhao** [comment]
I accept. Thanks!

**Comment by calebamiles** [comment]
I went ahead and excluded `docs/api-reference/autoscaling` as a path to consider; thanks for the pointer @grodrigues3. How do things look with the updated commits? I'm not sure what to think of the GCI e2e failures but they seem unrelated to the PR itself

**Comment by calebamiles** [comment]
@grodrigues3, had to prune the OWNERS files again to catch some late agreements. To keep the bot happy should I simply exclude the `docs` directory entirely since it applied a `do-not-merge` label _after_ your `LGTM`

**Comment by bgrant0607** [comment]
Avoid docs user-guide, admin, and getting-started

**Comment by grodrigues3** [comment]
@calebamiles the admin, getting-started-guides, user-guides directories have all been migrated to the docs repo and we PRs that modify those directories get do-not-merge labels.   can we delete all owners files from those subdirs?

**Comment by bgrant0607** [comment]
@calebamiles @grodrigues3 We're going to have to add people back to reviewers.

**Comment by calebamiles** [comment]
I've skipped the `docs` directory altogether, @bgrant0607. If desired I can migrate the `OWNERS` files within the `docs` directory as part of a separate PR. @grodrigues3 + @lavalamp feel free to take a look when you get a chance. Thanks!

**Comment by bgrant0607** [comment]
@calebamiles I have more context on who's who, the history of the code, etc. I'll make a pass over the PR and send you a patch or something.

**Comment by calebamiles** [comment]
Thanks, @bgrant0607, if you could actually make a PR against https://github.com/calebamiles/kubernetes-reviewers-configuration that would make my life super easy. I've been using that yaml doc to keep track of everyone's review preferences.

**Comment by bgrant0607** [comment]
@calebamiles What tool consumes that file?

**Comment by grodrigues3** [comment]
@calebamiles, blunderbuss also needs to be updated to read the new OWNERs format first right?

**Comment by bgrant0607** [comment]
@calebamiles 

When I looked at some of the OWNERS files, I noticed that findowners appeared to weigh recent PRs too heavily. One case in particular that stood out was joe2far, who was listed in dozens of OWNERS files due to a handful of mechanical PRs that touched lots of files. I appreciate that he accepted, but we want to choose reviewers with some background with the code they'd be reviewing.

**Comment by bgrant0607** [comment]
@calebamiles @grodrigues3 I don't think a simple git-based extractor is going to work. We're going to have to choose owners by hand. A tool that automatically extracted and analyzed the git history (LOC changed, #PRs, recency of PRs, ignore PRs that perform mechanical changes across the codebase, find previous PR reviewers) would be very useful raw data, though. 

I would like all of the mechanism changes to be ready, though.

Yes, the PR assigner needs to handle assignees AND reviewers.

Is the approval mechanism ready?

**Comment by bgrant0607** [comment]
I reverted the last commit, which removed reviewers, and then looked through and updated a few dozen OWNERS files (all in /test except e2e and e2e_node):
[owners-patch.txt](https://github.com/kubernetes/kubernetes/files/532471/owners-patch.txt)

We need to take into account:
* Reviewers (which isn't always the assignee, sadly) of relevant PRs, not just authors of the PRs
* [Area ownership](https://github.com/kubernetes/kubernetes/wiki)
* Continued activity on the project, not just on the particular file
* Meaningful vs mechanical changes (the latter often change isolated lines on files from many directories)
* Code moves that didn't preserve change history (e.g., from other repos)
* Generated code (e.g., /staging)
* Whether the directory is a leaf or not

I don't see easy ways to address the above programmatically.

**Comment by apelisse** [comment]
This is on-going, it just takes time!

**Comment by bgrant0607** [comment]
@apelisse Do we still plan to use this PR?

**Comment by apelisse** [comment]
@bgrant0607 We use this as the base PR for all other PRs. We could close certainly close it now though.

**Comment by rata** [comment]
@bgrant0607 @apelisse I'm quite excited/anxious to be a reviewer. Can I ask what is the plan? I've tried to follow, but it is not clear to me

**Comment by grodrigues3** [comment]
@rata Glad you're excited to be a reviewer.  The plan is to have a two tiered review process.  The first requires reviewers (someone like yourself) who has committed and reviewed some code.  They will be asked to read new PRs, comment on the design, style, overall quality of the code.  When a reviewer is satisfied, they may LGTM the code.  The second step is getting an approval from an Approver.  Approvers have generally reviewed/and or written a lot of code, and have a lot of experience with a large area in the project.  They can approve individual files which will eventually lead to a "approved" label and the PR getting merged.  This process will allow the project to grow our reviewer pool and eventually promote reviewers into approvers, allowing k8s reviews to keep pace with the code contribution.

**Comment by rata** [comment]
@grodrigues3 awesome! Thanks for the info! :)

**Comment by calebamiles** [comment]
Closing as subsumed by the work of @apelisse

**Pull request closed without merging**